### PR TITLE
Object3D: added attachTo() method

### LINKED ISSUE #6426: THREE Group and Object3D add child BUG.
Ricardo, why did you delete the reported issue?
I'm a 3DS Max user since the DOS era, so I know very well what I'm saying.
Again:
- import an object (OBJ, DAE, whatever) with many childrens.
- perform some scale/rotation/position changes on the imported object (all the childrens will be afected).
- detach one children (the mesh)
- create a new GROUP
- attach that mesh to the group.

Result: everything is messed up !
That previous child object will get position and rotation (0,0,0) and scale(1,1,1).
And that's wrong!
Perform the same operations with 3DS Max.
The new group (parent) will not mess up the new children parameters.

cheers

### PR DESCRIPTION
**Pull request by WestLangley**
`Object3D.attachTo( object )` adds this object as a child of the given object, while maintaining this object's world transform.

Fixes #6426.

This method alleviates the need for `THREE.SceneUtils.attach/detach()`, which can be deprecated in a later PR.

### CODE REVIEW

**Comment by gkjohnson** [comment]
Awesome! I like the addition of a function like this.

Some of my immediate thoughts are that the argument differences between `attachTo` and `add` may cause a bit of confusion and feels inconsistent:

```js
parent.add(child);

// vs

child.attachTo(parent);
```

And I know the name came from the `SceneUtils` class but I wonder if there might be a better word than `attach` to describe what the function does. Admittedly I'm having trouble coming up with alternatives, though.

**Comment by WestLangley** [comment]
The purpose of this method is to assign a new _parent_ while maintaining the object's world transform -- hence the API `attachTo( parent )`.

**Comment by gkjohnson** [comment]
I understand the purpose of the function but the description of the function could be changed to "add a child to an object while maintaining the child's world transform" in which case the api might look like `attach( child )`. Both would have the same outcome.

I'm just suggesting that an API be considered that is consistent with another function that does something very similar.

**Comment by Mugen87** [comment]
Should we also add `Object3D.detachFrom( object )`?

**Comment by WestLangley** [comment]
>Should we also add Object3D.detachFrom( object )?

What would be the purpose of such a method?

**Comment by Mugen87** [comment]
Well, `Object3D.detach()` would be sufficient. It would detach an object from the its parent and apply the current world transform to it.

**Comment by WestLangley** [comment]
You described what the method would _do_. I was asking for the _purpose_ of such a method.

**Comment by Mugen87** [comment]
>  I was asking for the purpose of such a method.

Isn't that obvious?^^

**Comment by gkjohnson** [comment]
@Mugen87 I was also thinking about whether or not a `detach` function might be useful but when would you want to use `detach` instead of `remove`?

**Comment by yomboprime** [comment]
> @Mugen87 I was also thinking about whether or not a `detach` function might be useful but when would you want to use `detach` instead of `remove`?

I guess one use case would be to let an object "fall" from its parent in a seamless animation.

It is the symmetric operation of `attachTo`

**Comment by WestLangley** [comment]
>I guess one use case would be to let an object "fall off" from its parent in a seamless animation.

How do you render an object that has a `null` parent?

**Comment by WestLangley** [comment]
>I was asking for the purpose of such a method.
>>Isn't that obvious?^^

Not to me. That is why I asked. Can you please answer my question, anyway?

**Comment by EliasHasle** [comment]
> > I guess one use case would be to let an object "fall off" from its parent in a seamless animation.
> 
> How do you render an object that has a null parent?

@WestLangley Would your solution for such "detachment" be to _attach_ the object to the scene (or physics container, if that is different in the application)? If so, would it be wrong to (for intuitive symmetry) have a `detach` method that attaches the object to the farthest ancestor object or to the first `Scene` object found among ancestors?

**Comment by fernandojsg** [comment]
If I understood you correctly `attachTo` and `detach` could be used for example to replace the current code for the WebVR dragging example code https://github.com/mrdoob/three.js/blob/dev/examples/webvr_dragging.html#L172-L210 so:
- Grab an object you `attachTo` the controller (=converting the world coordinates from the object to local on the controller)
- Release it you `detach` it from controller and add it to (=converting local coordinates for the object to world coordinates) or `attachTo` `scene` instead so you can skip the `detach` call.

**Comment by yomboprime** [comment]
> > I guess one use case would be to let an object "fall off" from its parent in a seamless animation.
> 
> How do you render an object that has a `null` parent?

I understand. So, one would not `detach` an object. Instead, one would do `object.attachTo( scene )`.

**Comment by Mugen87** [comment]
> So, one would not detach an object.

My idea was to call `Object3D.detach` and then add it manually to the scene graph. The purpose of the method is to maintain the world transform even if it's no part of an object hierarchy anymore.

**Comment by EliasHasle** [comment]
@Mugen87 That makes sense when juggling multiple scenes, I think.

**Comment by WestLangley** [comment]
>My idea was to call Object3D.detach and then add it manually to the scene graph.

That is what `object.attachTo( scene ) `is for.

**Comment by Mugen87** [comment]
Okay, I see. But I still think I would be useful to have to object in a detached state.

**Comment by EliasHasle** [comment]
> Okay, I see. But I still think I would be useful to have to object in a detached state (no parent) with the correct world transform.

    object.attachTo( scene );
    scene.remove( object );

`detach` may feel more natural. I may be slightly above average neurotic, but to me the sheer beauty of symmetry achieved by including `detach` makes it right. ;-)

**Comment by yomboprime** [comment]
> > Okay, I see. But I still think I would be useful to have to object in a detached state (no parent) with the correct world transform.
> 
> ```
> object.attachTo( scene );
> scene.remove( object );
> ```
> 
> `detach` may feel more natural. I may be slightly above average neurotic, but to me the sheer beauty of symmetry achieved by including `detach` makes it right. ;-)

I also find useful to convert an object to world space without scene. Also, find more beautiful that `attachTo` is its own inverse operation.

As `object.attachTo( object2 )` already is valid if object has no parent, and it is treated as being in world space. What about letting `object.attachTo( null )` be also valid? It just needs some minor changes, something like this:

```
attachTo: function () {

	// adds this object as a child of the given object, while maintaining this object's world transform

	var m = new Matrix4();

	return function attachTo( object ) {

		if ( object === null ) {

			m.identity();

		}
		else {

			object.updateWorldMatrix( true, false );

			m.getInverse( object.matrixWorld );

		}

		if ( this.parent !== null ) {

			this.parent.updateWorldMatrix( true, false );

			m.multiply( this.parent.matrixWorld );

		}

		this.applyMatrix( m );

		this.updateWorldMatrix( false, false );

		if ( object !== null ) {

			object.add( this );

		}
		else if ( this.parent !== null ) {

			object.parent.remove( object );

		}

		return this;

	};

}(),
```

**Comment by WestLangley** [comment]
`Object3D.attachTo()` was designed to generalize and replace both `THREE.SceneUtils.attach()` and `THREE.SceneUtils.detach()`. 

```js
object.attachTo( newParent ); // detaches from oldParent...

object.attachTo( oldParent ); // ...now it's back where it started
```

**Review by Mugen87 — APPROVED** [review]

**Comment by mrdoob** [comment]
I like this but I feel it's backwards.

Couldn't it be...?

```js
parent.attach( object );
scene.attach( object );
```

**Comment by mrdoob** [comment]
I think adding `attach()` and `detach()` to `Object3D` seems more handy. I can see situations where you may want to detach an object without keeping it in the scene (but maintaining its world matrix) and then attaching it again to a different parent.

Having both, `add()` and `attach()` use the same pattern `parent.method( child )` seems also easy to explain to users.

**Comment by WestLangley** [comment]
>I like this but I feel it's backwards.
>>Couldn't it be...?
>>>parent.attach( object );
>>>scene.attach( object );

>Having both, add() and attach() using the same pattern parent.method( child ) seems also easier to explain to users.

I think users will be confused with the subtle difference between `object.add( child )` and `object.attach( child )`.

> I can see situations where you may want to detach an object without keeping it in the scene (but maintaining its world matrix)

Really? What would be your examples? We have millions of users and not one has requested a special method to do that.

In the rare case someone does want to do that...
```js
child.matrix.copy( child.matrixWorld );
parent.remove( child );
```

**Comment by WestLangley** [comment]
Closing in favor of #16528.
